gift-certificate-application/service/src/test/java/com/epam/esm/service/impl/GiftCertificateServiceImplTest.java

```diff
@@ -0,0 +1,301 @@
+        GiftCertificate actualResult = giftCertificateService.create(giftCertificate);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testCreateShouldThrowFieldValidationExceptionWhenTagInvalid() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(false);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(true);
+        assertThrows(FieldValidationException.class, () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testCreateShouldThrowFieldValidationExceptionWhenGiftCertificateInvalid() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(true);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(false);
+        assertThrows(FieldValidationException.class, () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testCreateShouldThrowGiftCertificateAlreadyExistException() {
+        when(tagValidator.isNameValid(anyString())).thenReturn(true);
+        when(giftCertificateValidator.validateAll(anyString(), anyString(), any(), anyShort())).thenReturn(true);
+        when(giftCertificateDao.findByName(anyString())).thenReturn(Optional.of(giftCertificate));
+        assertThrows(GiftCertificateAlreadyExistException.class,
+                () -> giftCertificateService.create(giftCertificate));
+    }
+
+    @Test
+    void testFindAll() {
+        List<GiftCertificate> expectedResult = giftCertificateList;
+        when(giftCertificateDao.findAll(pagination, GiftCertificate.class)).thenReturn(giftCertificateList);
+        when(giftCertificateDao.findEntitiesNumber(GiftCertificate.class)).thenReturn(ENTITIES_NUMBER);
+        when(paginationValidator.validatePagination(pagination, ENTITIES_NUMBER)).thenReturn(pagination);
+        List<GiftCertificate> actualResult = giftCertificateService.findAll(pagination);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testFindById() {
+        GiftCertificate expectedResult = giftCertificate;
+        when(giftCertificateDao.findById(GIFT_CERTIFICATE_ID, GiftCertificate.class))
+                .thenReturn(Optional.ofNullable(giftCertificate));
+        GiftCertificate actualResult = giftCertificateService
+                .findByGiftCertificateId(GIFT_CERTIFICATE_ID);
+        assertEquals(expectedResult, actualResult);
+    }
+
+    @Test
+    void testFindByIdShouldThrowGiftCertificateNotFoundException() {
+        when(giftCertificateDao.findById(GIFT_CERTIFICATE_ID, GiftCertificate.class)).thenReturn(Optional.empty());
+        assertThrows(GiftCertificateNotFoundException.class, () -> giftCertificateService
+                .findByGiftCertificateId(tag.getId()));
+    }
+
+    @Test
+    void testFindByAttributes() {
+        List<GiftCertificate> expectedResult = giftCertificateList;
```

Comment:
I suppose you can directly use giftCertificateList and do not create an additional variable